### PR TITLE
M2m2r milestoning support

### PR DESCRIPTION
**Pull request by PrateekGarg-gs**
#### What type of PR is this?

Imporvement

#### What does this PR do / why is it needed ?

This Pr adds support for milestoning in m2m2r cases
- change scan properties for milestoned qualified property
- add support for milestoned source to non-milestoned target in m2m via no-arg property
- propagate parent Milestoning context for no arg properties in relational graph fetch
- minor refactor

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/modelToModelToRelational/milestoned_graphFetch_m2m2r.pure [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        }
+      }#
+    )
+    ->meta::pure::mapping::from($mapping,$runtime)
+    };
+   let result = meta::legend::executeLegendQuery($query, [], ^meta::pure::runtime::ExecutionContext(), meta::relational::extension::relationalExtensions());
+   assertJsonStringsEqual('{"builder":{"_type":"json"},"values":[{"id":2,"name":"ProductName2","synonyms(2020-10-15T00:00:00+0000)":[{"synonym":"GS-Mod-S1"},{"synonym":"GS-Mod-S3"}]},{"id":3,"name":"ProductName3","synonyms(2020-10-15T00:00:00+0000)":[{"synonym":"GS-Mod-S3"}]}]}', $result);
+}
+
+function <<test.Test, test.AlloyOnly, test.ToFix>>  meta::pure::graphFetch::tests::m2m2r::milestoning::toMilestonedTargetProperty::testToMilestonedTargetProperty_variableDate():Boolean[1]
```
Comment (PrateekGarg-gs) [inline legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/modelToModelToRelational/milestoned_graphFetch_m2m2r.pure:41]: this test needs fixing of date variable handling in relational graph fetch . will fix in upcoming PR with some other changes

File: legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/modelToModelToRelational/milestoned_graphFetch_m2m2r.pure [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+Mapping meta::relational::tests::m2m2r::milestoning::toMilestonedTargetProperty::TargetToModelMappingWithMilestonedComplexProperty
+(
+  include meta::relational::tests::milestoning::milestoningmap
+
+  *meta::relational::tests::milestoning::TargetProduct1: Pure
+  {
+    ~src meta::relational::tests::milestoning::Product
+    id: $src.id,
+    name: $src.name,
+    synonymsAllVersions: $src.synonymsAllVersions
```
Comment (rafaelbey) [inline legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/modelToModelToRelational/milestoned_graphFetch_m2m2r.pure:191]: I wonder if there are better ways to do this?  Seems confusing :(

Comment (PrateekGarg-gs) [reply legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/modelToModelToRelational/milestoned_graphFetch_m2m2r.pure:191]: should be  synonyms: $src.synonyms ?

Comment (rafaelbey) [reply legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/modelToModelToRelational/milestoned_graphFetch_m2m2r.pure:191]: Yeah, I wonder if from the Studio grammar perspective this is more understandable to users.  We can internally convert to 'AllVersions" if requried.

Comment (PrateekGarg-gs) [reply legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/modelToModelToRelational/milestoned_graphFetch_m2m2r.pure:191]: Yes that seems more intuitive . we can support both patterns. it needs a fix in m2m source tree calculation

File: legend-engine-pure/legend-engine-pure-code/legend-engine-pure-code-compiled-core/src/main/resources/core/pure/lineage/scanProperties.pure [hunk 3]
```diff
@@ -467,17 +468,36 @@ function <<access.private>> meta::pure::lineage::scanProperties::processQualifie
...
+       | ^PropertyPathNode(
                           class                 = $srcClass,
-                          property              = $q,
-                          nestedQualifier       = $qualifierBody.result,
-                          parameters            = $parameters,
-                          nestedQualifierReturn = $qualifierBody->last().current
-                       );
+                          property              = if(hasSufficientValuesForPropertyParameters( $q,$parameters),
+                                                       | $q,
+                                                       | $q->switchToNoArgMilestonedGeneratedQualifiedProperty()->toOne()
```
[the remark this answers is not in the document]
Comment (PrateekGarg-gs) [reply legend-engine-pure/legend-engine-pure-code/legend-engine-pure-code-compiled-core/src/main/resources/core/pure/lineage/scanProperties.pure:477]: yes this is the exact case being handled with this MR. 
consider this flattening propery m2m mapping   > **synonymNames:  $src.synonyms.synonym**
Here rhs $src.synonyms expects a date but we arent supplying it as we want to infer Milestoning context from source root. 
Note - there is no way user can specify the date var used at query time into mapping.

File: legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/modelToModelToRelational/milestoned_graphFetch_m2m2r.pure [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+          id,
+          name,
+          synonyms(%2020-10-15T00:00:00)
+          {synonym}
+        }
+      }#
+    )
+    ->meta::pure::mapping::from($mapping,$runtime)
+    };
+   let result = meta::legend::executeLegendQuery($query, [], ^meta::pure::runtime::ExecutionContext(), meta::relational::extension::relationalExtensions());
```
[the remark this answers is not in the document]
Comment (PrateekGarg-gs) [reply legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/modelToModelToRelational/milestoned_graphFetch_m2m2r.pure:37]: actually executeLegendQuery only returns results not the sql run ( like activities in old flow) I can add another statement to generate plan before execute. 

But adding sql assertions is **not very maintainable** as sqls keep on changing with db versions and our translation changes . End to end behaviour should be same from engine perspective ?

File: legend-engine-pure/legend-engine-pure-code/legend-engine-pure-code-compiled-core/src/main/resources/core/pure/milestoning/milestoning.pure [hunk 5]
```diff
@@ -85,4 +85,28 @@ function meta::pure::milestoning::isBusinessTemporal(type:Type[1]):Boolean[1]
 function meta::pure::milestoning::isBiTemporal(type:Type[1]):Boolean[1]
 {
    ['bitemporal']->exists(s| $type->hasStereotype($s, temporal))
+}
+
+
+function meta::pure::milestoning::isMilestonedGeneratedQualifiedProperty(p:AbstractProperty<Any>[1]):Boolean[1]{
```
[the remark this answers is not in the document]
Comment (PrateekGarg-gs) [reply legend-engine-pure/legend-engine-pure-code/legend-engine-pure-code-compiled-core/src/main/resources/core/pure/milestoning/milestoning.pure:91]: this file contains these functions -isMilestonedGeneratedDateProperty ,  but yes these functions are all over the place , can be refactored strategically

File: legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/graphFetch/relationalGraphFetch.pure [hunk 6]
```diff
@@ -982,15 +983,22 @@ function <<access.private>> meta::relational::graphFetch::executionPlan::newVars
...
-   if($isQualified,
-      | let dummyL = {|1+2};
-        let dummySFE = $dummyL.expressionSequence->cast(@SimpleFunctionExpression)->map(x | ^$x(func = $propTree.property, genericType = $propTree.property.genericType, parametersValues=[^InstanceValue(genericType=^GenericType(rawType=$parentTree->typeFromGraphFetchTree()), multiplicity=ZeroMany, values = [])]->concatenate($propTree.property->cast(@QualifiedProperty<Any>)->functionType().parameters->evaluateAndDeactivate()->tail())))->toOne();
-        getMilestoningContextForQualifiedProperty($dummySFE->evaluateAndDeactivate(), [], [], $state, ^Map<VariableExpression, ValueSpecification>(), noDebug(), $extensions);,
-      | []
-   );
+   let isQualifiedMilestoned = $propTree.property->instanceOf(QualifiedProperty) && $propTree.property->hasGeneratedMilestoningPropertyStereotype();
+   if($isQualifiedMilestoned ,
+    |  if($propTree.property->isNoArgMilestonedGeneratedQualifiedProperty(),
+        | assert($parentMilestoningContext->isNotEmpty(),'No-Arg Milestoned property >'+ $propTree.property.name->toOne()+' should be used in a Milestoning Context');
```
[the remark this answers is not in the document]
Comment (PrateekGarg-gs) [reply legend-engine-xts-relationalStore/legend-engine-xt-relationalStore-generation/legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/graphFetch/relationalGraphFetch.pure:991]: apparantly there was support for milestoning context propgation was at compiler level where dates were populated, vut not at relational store level

File: legend-engine-pure/legend-engine-pure-code/legend-engine-pure-code-compiled-core/src/main/resources/core/pure/lineage/scanProperties.pure (hunk 3 above)
[the remark this answers is not in the document]
Comment (PrateekGarg-gs) [reply legend-engine-pure/legend-engine-pure-code/legend-engine-pure-code-compiled-core/src/main/resources/core/pure/lineage/scanProperties.pure:477]: put a flag to explodeDateProperty

File: legend-engine-pure/legend-engine-pure-code/legend-engine-pure-code-compiled-core/src/main/resources/core/pure/lineage/scanProperties.pure [hunk 7]
```diff
@@ -467,17 +468,36 @@ function <<access.private>> meta::pure::lineage::scanProperties::processQualifie
...
+                                nestedQualifierReturn = $qualifierBody->last().current
+                            );
+     );
    if($append,|^$list(values = $list.values->concatenate($property)),|^$list(values = $property->concatenate($list.values)));
 }
 
+
+function <<access.private>> meta::pure::lineage::scanProperties::hasSufficientValuesForPropertyParameters(q:QualifiedProperty<Any>[1], parameters:ValueSpecification[*]):Boolean[1]
+{
+   $q->functionType().parameters->evaluateAndDeactivate()->size() - ($parameters->size()) == 1;
```
[the remark this answers is not in the document]
Comment (PrateekGarg-gs) [reply legend-engine-pure/legend-engine-pure-code/legend-engine-pure-code-compiled-core/src/main/resources/core/pure/lineage/scanProperties.pure:498]: ->tail()

**Review by davidharte-gs — APPROVED** [review]

**Pull request merged**